# 1.- Librerias necesarias

In [1]:
#Importo librerias
# cargo librerías necesarias
import numpy  as np  
import pandas as pd
from datetime import datetime
from sklearn.preprocessing import StandardScaler, MinMaxScaler
from sklearn.model_selection import train_test_split
from sklearn.metrics import explained_variance_score, mean_squared_error, r2_score


import time

from keras import models, layers, optimizers, regularizers
from keras.utils.vis_utils import model_to_dot
from IPython.display import SVG

import matplotlib.pyplot as plt # para dibujar
%matplotlib inline

import warnings
warnings.filterwarnings('ignore')

Using TensorFlow backend.


Ruta en google drive del proyecto

In [0]:
ruta = "/content/gdrive/My Drive/KeepCoding/ProyectoFinal/"

# 2.- Carga de datos train y test

In [3]:
from google.colab import drive
drive.mount('/content/gdrive')

Go to this URL in a browser: https://accounts.google.com/o/oauth2/auth?client_id=947318989803-6bn6qk8qdgf4n4g3pfee6491hc0brc4i.apps.googleusercontent.com&redirect_uri=urn%3aietf%3awg%3aoauth%3a2.0%3aoob&response_type=code&scope=email%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdocs.test%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive.photos.readonly%20https%3a%2f%2fwww.googleapis.com%2fauth%2fpeopleapi.readonly

Enter your authorization code:
··········
Mounted at /content/gdrive


Cargo los datos para train

In [6]:
path_csv_train = ruta + 'data/listings_train.csv'
house_madrid_train = pd.read_csv(path_csv_train, sep =',', decimal='.') # cargamos fichero
print(f"El dataset contiene {len(house_madrid_train)} viviendas Airbnb con {len(house_madrid_train.columns)} características")
pd.set_option('display.max_columns', len(house_madrid_train)) # Para ver todas las columnas
house_madrid_train.head(5)  

El dataset contiene 9060 viviendas Airbnb con 219 características


Unnamed: 0     Price        id  host_is_superhost  host_listings_count  \
0           0  3.091042  38356185                0.0             0.693147   
1           1  4.077537  30488431                0.0             4.143135   
2           2  4.248495  35532108                0.0             2.639057   
3           3  4.060443  16451687                0.0             4.356709   
4           4  4.248495  29274213                1.0             0.000000   

   host_identity_verified  is_location_exact  accommodates  bathrooms  \
0                     0.0                  0      0.000000        0.0   
1                     0.0                  1      1.098612        0.0   
2                     0.0                  1      1.098612        0.0   
3                     0.0                  1      1.386294        0.0   
4                     0.0                  1      1.386294        0.0   

   security_deposit  cleaning_fee  extra_people  minimum_nights  \
0          -4.60517      2.995732      1.609438        0.000000   
1           4.60517      3.688879      2.302585        0.000000   
2          -4.60517     -4.605170     -4.605170        0.000000   
3          -4.60517      3.218876     -4.605170        0.000000   
4           4.60517      3.912023      2.484907        1.098612   

   maximum_nights  availability_90  number_of_reviews  instant_bookable  \
0        7.025538               58           2.079442                 1   
1        7.025538               21           2.708050                 1   
2        7.025538               63           1.791759                 0   
3        3.332205               64           5.023881                 1   
4        6.907755               75           3.367296                 1   

   sentiment  Topic_general  Topic_descripcion  Topic_atracciones  \
0   0.156667      -0.407397          -3.282160           0.013976   
1   0.287834      -4.605170          -0.976263           0.144895   
2   0.383571      -4.605170          -1.396565           0.238780   
3   0.395745      -4.605170          -4.605170           0.087998   
4   0.363727      -4.605170          -1.662112           0.021387   

   Topic_servicios  Topic_transporte  Dias_activo  Internet  \
0        -2.791681         -1.506077        846.0       0.0   
1        -0.856820         -2.946893       1733.0       1.0   
2        -2.571690         -0.847858        170.0       0.0   
3        -0.121937         -4.136825       1059.0       1.0   
4        -0.828728         -1.047018        807.0       0.0   

   Free street parking  Oven  Crib  Cooking basics  Luggage dropoff allowed  \
0                  0.0   0.0   0.0             0.0                      0.0   
1                  0.0   0.0   1.0             1.0                      1.0   
2                  0.0   0.0   0.0             0.0                      0.0   
3                  0.0   0.0   1.0             1.0                      0.0   
4                  0.0   1.0   0.0             1.0                      0.0   

   Smoke detector  Room-darkening shades  Dishwasher  \
0             0.0                    0.0         0.0   
1             0.0                    0.0         0.0   
2             0.0                    0.0         0.0   
3             1.0                    0.0         0.0   
4             0.0                    0.0         0.0   

   Paid parking on premises  Wide entrance  No stairs or steps to enter  \
0                       0.0            0.0                          1.0   
1                       0.0            0.0                          0.0   
2                       0.0            0.0                          0.0   
3                       0.0            1.0                          0.0   
4                       0.0            0.0                          1.0   

   Cable TV  Paid parking off premises  Elevator  Buzzer/wireless intercom  \
0       0.0                        0.0       0.0                       0.0   
1       0.0                       

Cargo los datos para test

In [7]:
path_csv_test = ruta + 'data/listings_test.csv'
house_madrid_test = pd.read_csv(path_csv_test, sep =',', decimal='.') # cargamos fichero
print(f"El dataset contiene {len(house_madrid_test)} viviendas Airbnb con {len(house_madrid_test.columns)} características")
pd.set_option('display.max_columns', len(house_madrid_test)) # Para ver todas las columnas
house_madrid_test.head(5) 

El dataset contiene 3872 viviendas Airbnb con 219 características


Unnamed: 0     Price        id  host_is_superhost  host_listings_count  \
0           0  4.584967  16064559                0.0             1.791759   
1           2  4.499810  13434343                0.0             1.386294   
2           3  4.174387  11645913                1.0             0.693147   
3           4  4.382027  32906611                1.0             0.000000   
4           5  4.317488  16356985                0.0             0.693147   

   host_identity_verified  is_location_exact  accommodates  bathrooms  \
0                     1.0                  1      1.791759        0.0   
1                     1.0                  1      1.609438        0.0   
2                     0.0                  1      0.693147        0.0   
3                     0.0                  1      1.386294        0.0   
4                     0.0                  1      1.386294        0.0   

   security_deposit  cleaning_fee  extra_people  minimum_nights  \
0         -4.605170      3.555348     -4.605170        0.693147   
1          4.787492      3.688879      3.688879        0.693147   
2          4.605170      2.833213      2.995732        1.098612   
3          5.298317      4.043051      2.302585        0.693147   
4          5.010635      3.465736      2.708050        0.693147   

   maximum_nights  availability_90  number_of_reviews  instant_bookable  \
0        5.991465               63           4.532599                 1   
1        7.025538               72           1.386294                 1   
2        4.605170               55           5.214936                 1   
3        3.806662               62           3.367296                 0   
4        6.551080               50           4.624973                 1   

   sentiment  Topic_general  Topic_descripcion  Topic_atracciones  \
0   0.415570       -4.60517          -0.755703           0.320701   
1   0.480752       -4.60517          -2.589536           0.780409   
2   0.467749       -4.60517          -4.114707           0.284692   
3   0.386355       -4.60517          -0.738213           0.159290   
4   0.426225       -4.60517          -1.577694           0.319856   

   Topic_servicios  Topic_transporte  Dias_activo  Internet  \
0        -2.243319         -2.283258       1706.0       1.0   
1        -3.075842         -2.371856       1288.0       1.0   
2        -0.862586         -1.289782       1801.0       1.0   
3        -1.185480         -2.886151        246.0       0.0   
4        -0.796729         -3.855013       1061.0       0.0   

   Free street parking  Oven  Crib  Cooking basics  Luggage dropoff allowed  \
0                  0.0   1.0   1.0             1.0                      1.0   
1                  0.0   1.0   0.0             1.0                      0.0   
2                  0.0   1.0   0.0             1.0                      1.0   
3                  0.0   1.0   0.0             1.0                      0.0   
4                  0.0   0.0   0.0             0.0                      0.0   

   Smoke detector  Room-darkening shades  Dishwasher  \
0             0.0                    0.0         1.0   
1             0.0                    0.0         0.0   
2             1.0                    0.0         1.0   
3             1.0                    0.0         1.0   
4             0.0                    0.0         0.0   

   Paid parking on premises  Wide entrance  No stairs or steps to enter  \
0                       0.0            0.0                          0.0   
1                       0.0            0.0                          0.0   
2                       1.0            0.0                          0.0   
3                       0.0            0.0                          0.0   
4                       0.0            0.0                          0.0   

   Cable TV  Paid parking off premises  Elevator  Buzzer/wireless intercom  \
0       1.0                        1.0       0.0                       0.0   
1       0.0                       

Elimino las columnas Unname

In [0]:
house_madrid_train.drop(house_madrid_train.columns[house_madrid_train.columns.str.contains('Unnamed',case = False)],axis = 1, inplace = True)
house_madrid_test.drop(house_madrid_test.columns[house_madrid_test.columns.str.contains('Unnamed',case = False)],axis = 1, inplace = True)

Elimino las columnas "id"

In [0]:
house_madrid_train.drop ('id', axis=1, inplace=True)
house_madrid_test.drop ('id', axis=1, inplace=True)

#3.- Separación X e Y. Normalización

Train

In [0]:
# Separamos X and y
X_train = house_madrid_train.drop('Price', axis=1)
y_train = house_madrid_train.Price

# Scaling
scaler = StandardScaler()
X_train = pd.DataFrame(scaler.fit_transform(X_train), columns=list(X_train.columns))

Test

In [0]:
# Separamos X and y
X_test = house_madrid_test.drop('Price', axis=1)
y_test = house_madrid_test.Price

# Scaling
scaler = StandardScaler()
X_test = pd.DataFrame(scaler.fit_transform(X_test), columns=list(X_test.columns)) 

# 4.- Modelado 

Voy a aplicar diferentes modelos "Supervised Deep Learning" y los compararemos entre si. La métrica que usaré al ser un problema de regresión es el RMSE (Error cuadrático medio)





Al haber dividido en train y test antes de cualquier procesado y haber realizado la conversión de variables categóricas con el metodo dumy podría ocurrir que tenga diferente número de características en un cojunto y en otro. Voya asegurarme que tengo en los dos lo mismo y que las características están ordenadas de la misma manera:

In [0]:
missing_cols = set (X_train.columns) - set (X_test.columns)
#Caso de que que en test no tenga las mismas columnas las añado con valor 0
for c in missing_cols:
    X_test [c]=0
#Me aseguro que el orden de las columnas en test sea el mismo que en train
X_test = X_test[X_train.columns]

In [13]:
X_train.head(1)

host_is_superhost  host_listings_count  host_identity_verified  \
0          -0.606767            -0.201163               -0.756191   

   is_location_exact  accommodates  bathrooms  security_deposit  cleaning_fee  \
0          -1.561515     -1.862962  -0.360777         -0.906197      0.542656   

   extra_people  minimum_nights  maximum_nights  availability_90  \
0      0.592868       -1.122311        0.714385         0.608625   

   number_of_reviews  instant_bookable  sentiment  Topic_general  \
0           -0.72465          0.780719  -1.664477       3.281426   

   Topic_descripcion  Topic_atracciones  Topic_servicios  Topic_transporte  \
0          -0.652486          -1.029992        -0.713435          0.122193   

   Dias_activo  Internet  Free street parking     Oven      Crib  \
0    -0.713977 -0.552698            -0.392045 -0.73879 -0.381295   

   Cooking basics  Luggage dropoff allowed  Smoke detector  \
0         -0.9623                 -0.53839       -0.522673   

   Room-darkening shades  Dishwasher  Paid parking on premises  Wide entrance  \
0              -0.345467    -0.57888                 -0.491361      -0.392232   

   No stairs or steps to enter  Cable TV  Paid parking off premises  Elevator  \
0                     2.086012 -0.347266                  -0.624156 -1.142516   

   Buzzer/wireless intercom  Extra pillows and blankets  Bed linens  \
0                 -0.480254                   -0.672425   -0.974067   

   Host greets you  Family/kid friendly  Fire extinguisher   Hangers  \
0        -0.873689            -0.783101          -0.576161  0.378822   

   Patio or balcony  Microwave  Pets allowed  Smoking allowed        TV  \
0         -0.415893    -1.0769      2.536202         2.004149  0.476944   

     Washer  Shampoo  Hair dryer    Dryer  Long term stays allowed  Breakfast  \
0  0.440086 -1.83302   -1.983593 -0.46768                -0.731842  -0.361668   

   Coffee maker  First aid kit      Iron  Private entrance  \
0     -0.993619      -0.648294  0.533612         -0.390732   

   Laptop friendly workspace  Lock on bedroom door  Hot water  \
0                  -1.436273             -0.412968   0.617672   

   Air conditioning     Stove  Refrigerator  \
0          0.719436 -0.794702     -1.173065   

   host_response_time_a few days or more  host_response_time_unknown  \
0                              -0.112383                   -0.359518   

   host_response_time_within a day  host_response_time_within a few hours  \
0                        -0.257164                              -0.378822   

   host_response_time_within an hour  host_response_rate_100%  \
0                           0.677276                  0.72971   

   host_response_rate_50-89%  host_response_rate_90-99%  \
0                  -0.348064                  -0.346667   

   neighbourhood_cleansed_Acacias  neighbourhood_cleansed_Argüelles  \
0                       -0.098467                         -0.122988   

   neighbourhood_cleansed_Cortes  neighbourhood_cleansed_Cuatro Caminos  \
0                      -0.231805                              -0.098467   

   neighbourhood_cleansed_Delicias  neighbourhood_cleansed_Embajadores  \
0                         9.612599                           -0.423714   

   neighbourhood_cleansed_Goya  neighbourhood_cleansed_Guindalera  \
0                     -0.11438                          -0.096155   

   neighbourhood_cleansed_Ibiza  neighbourhood_cleansed_Justicia  \
0                     -0.091975                        -0.257408   

   neighbourhood_cleansed_Numancia  neighbourhood_cleansed_Otro_Barrio  \
0                        -0.096155                           -0.658613   

   neighbourhood_cleansed_Pacífico  neighbourhood_cleansed_Palacio  \
0                        -0.109322                       -0.299109   

   neighbourhood_cleansed_Palos de Moguer  \
0                               -0.143179   

   neighbourhood_cleansed_Puerta del Angel  neighbourhood_cleansed_Recoletos  \

In [14]:
X_test.head(1)

host_is_superhost  host_listings_count  host_identity_verified  \
0          -0.600421             0.516655                1.292418   

   is_location_exact  accommodates  bathrooms  security_deposit  cleaning_fee  \
0           0.636745      1.324146  -0.407383         -0.901178      0.706543   

   extra_people  minimum_nights  maximum_nights  availability_90  \
0     -1.137171        0.104546        0.160193         0.782578   

   number_of_reviews  instant_bookable  sentiment  Topic_general  \
0           0.922395          0.780583   0.230016      -0.387168   

   Topic_descripcion  Topic_atracciones  Topic_servicios  Topic_transporte  \
0           1.286738           0.410101        -0.244869         -0.566405   

   Dias_activo  Internet  Free street parking      Oven      Crib  \
0     0.400495  1.779866            -0.400838  1.305339  2.533303   

   Cooking basics  Luggage dropoff allowed  Smoke detector  \
0        1.004141                 1.857572        -0.52955   

   Room-darkening shades  Dishwasher  Paid parking on premises  Wide entrance  \
0              -0.364908    1.683349                 -0.490133      -0.409048   

   No stairs or steps to enter  Cable TV  Paid parking off premises  Elevator  \
0                    -0.486077  2.962731                   1.568513 -1.150708   

   Buzzer/wireless intercom  Extra pillows and blankets  Bed linens  \
0                 -0.480384                    1.452598    1.014569   

   Host greets you  Family/kid friendly  Fire extinguisher   Hangers  \
0         1.136276             1.280391          -0.576953  0.365362   

   Patio or balcony  Microwave  Pets allowed  Smoking allowed        TV  \
0          2.349153   0.891209      2.544583        -0.508702  0.489323   

     Washer   Shampoo  Hair dryer     Dryer  Long term stays allowed  \
0  0.434119  0.535942    0.500646  2.123164                 1.380908   

   Breakfast  Coffee maker  First aid kit      Iron  Private entrance  \
0  -0.350671      0.973492      -0.644364  0.541529         -0.401272   

   Laptop friendly workspace  Lock on bedroom door  Hot water  \
0                   0.679879             -0.421869   0.600421   

   Air conditioning     Stove  Refrigerator  \
0          0.728725  1.215437      0.826727   

   host_response_time_a few days or more  host_response_time_unknown  \
0                              -0.109649                   -0.363088   

   host_response_time_within a day  host_response_time_within a few hours  \
0                        -0.261596                              -0.380192   

   host_response_time_within an hour  host_response_rate_100%  \
0                           0.683943                -1.366025   

   host_response_rate_50-89%  host_response_rate_90-99%  \
0                  -0.354836                   2.937999   

   neighbourhood_cleansed_Acacias  neighbourhood_cleansed_Argüelles  \
0                        -0.11553                         -0.116673   

   neighbourhood_cleansed_Cortes  neighbourhood_cleansed_Cuatro Caminos  \
0                      -0.226532                              -0.091287   

   neighbourhood_cleansed_Delicias  neighbourhood_cleansed_Embajadores  \
0                        -0.102169                           -0.428222   

   neighbourhood_cleansed_Goya  neighbourhood_cleansed_Guindalera  \
0                    -0.103451                          -0.102169   

   neighbourhood_cleansed_Ibiza  neighbourhood_cleansed_Justicia  \
0                     -0.094121                        -0.253039   

   neighbourhood_cleansed_Numancia  neighbourhood_cleansed_Otro_Barrio  \
0                        -0.099556                           -0.689645   

   neighbourhood_cleansed_Pacífico  neighbourhood_cleansed_Palacio  \
0                        -0.096875                       -0.284005   

   neighbourhood_cleansed_Palos de Moguer  \
0                               -0.137649   

   neighbourhood_cleansed_Puerta del Angel  neighbourhood_cleansed_Recole

## 4.1. Modelo 1: Red neuronal con 3 capas
Voy a comenzar con una red neuronal de 3 capas. Utilizaré 3 capas densas, una fución de activación de tipo relu y para la salida utilizaré una función lineal al ser un problema de regresión. La función de pérdidas será la "mean squered_error'. Utilizaré además como optimizador adam con su lr por defecto

In [15]:
#Voy a comprobar el número de características que tengo para que la primera capa densa tenga más neuronas que este 
#número de características
len  (list(X_train.columns))

216

In [16]:
#Creo el modelo
modelo = models.Sequential()
modelo.add (layers.Dense (256, input_shape=(X_train.shape[1],), activation='relu'))
modelo.add(layers.Dense(256, activation='relu'))
modelo.add(layers.Dense(128, activation='relu'))
modelo.add(layers.Dense(1, activation='linear'))

#Compilo el modelo 
modelo.compile(loss='mean_squared_error',
               optimizer='adam',
               metrics=['mean_squared_error'])

print(modelo.summary())








Model: "sequential_1"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_1 (Dense)              (None, 256)               55552     
_________________________________________________________________
dense_2 (Dense)              (None, 256)               65792     
_________________________________________________________________
dense_3 (Dense)              (None, 128)               32896     
_________________________________________________________________
dense_4 (Dense)              (None, 1)                 129       
Total params: 154,369
Trainable params: 154,369
Non-trainable params: 0
_________________________________________________________________
None


In [17]:
#Entreno el modelo con 100 épocas y batch_size de 256
modelo_start = time.time()
modelo.fit(X_train,
                  y_train,
                  epochs=100,
                  batch_size=256,
                  validation_split = 0.1  #Con esto se separa una porción del conjunto de entratamiento y se usa como
                                          #validación, con esa porción no se entrena
                  )
modelo_end = time.time()
print(f"Tiempo de ejecución: {round((modelo_end - modelo_start)/60,1)} minutes")




Train on 8154 samples, validate on 906 samples
Epoch 1/100





8154/8154 [==============================] - 9s 1ms/step - loss: 2.0650 - mean_squared_error: 2.0650 - val_loss: 0.4619 - val_mean_squared_error: 0.4619
Epoch 2/100
8154/8154 [==============================] - 0s 15us/step - loss: 0.2908 - mean_squared_error: 0.2908 - val_loss: 0.2034 - val_mean_squared_error: 0.2034
Epoch 3/100
8154/8154 [==============================] - 0s 13us/step - loss: 0.1596 - mean_squared_error: 0.1596 - val_loss: 0.1694 - val_mean_squared_error: 0.1694
Epoch 4/100
8154/8154 [==============================] - 0s 14us/step - loss: 0.1236 - mean_squared_error: 0.1236 - val_loss: 0.1525 - val_mean_squared_error: 0.1525
Epoch 5/100
8154/8154 [==============================] - 0s 14us/step - loss: 0.1030 - mean_squared_error: 0.1030 - val_loss: 0.1452 - val_mean_squared_error: 0.1452
Epoch 6/100
8154/8154 [==============================] - 0s 16us/step - loss: 0.0873 - mean_squared_error: 0.0873 - 

In [18]:
#Evaluo el modelo con las muestras de test
y_test_pred = modelo.predict(X_test)
y_train_pred = modelo.predict(X_train)
print("Training MSE:", round(mean_squared_error(y_train, y_train_pred),4))
print("Validation MSE:", round(mean_squared_error(y_test, y_test_pred),4))
print("\nTraining r2:", round(r2_score(y_train, y_train_pred),4))
print("Validation r2:", round(r2_score(y_test, y_test_pred),4))

Training MSE: 0.0141
Validation MSE: 0.1448

Training r2: 0.9602
Validation r2: 0.5903


En este primer intento se ve que el resultado en train es mucho mejor que el que obtenermos en test lo cual indica que tenemos un problema de overfitting. Voy a aplicar regularización para intentar solucionar el problema

## 4.2. Modelo 2: Red neuronal con 4 capas, regularización L1 y más épocas
Para reducir el overfitting voy a usar la regularización L1. Añado además una capa adicional para ver si podemos mejorar el resultado (r2)

In [65]:
# Construyo el modelo
modelo2 = models.Sequential()
modelo2.add(layers.Dense(512, input_shape=(X_train.shape[1],), kernel_regularizer=regularizers.l1(0.005), activation='relu'))
modelo2.add(layers.Dense(256, kernel_regularizer=regularizers.l1(0.005), activation='relu'))
modelo2.add(layers.Dense(256, kernel_regularizer=regularizers.l1(0.005), activation='relu'))
modelo2.add(layers.Dense(128, kernel_regularizer=regularizers.l1(0.005), activation='relu'))
modelo2.add(layers.Dense(1, activation='linear'))

# Compilo el modelo
modelo2.compile(loss='mean_squared_error',
                optimizer='adam',
                metrics=['mean_squared_error'])

# Model summary
print(modelo2.summary())

Model: "sequential_17"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_80 (Dense)             (None, 512)               111104    
_________________________________________________________________
dense_81 (Dense)             (None, 256)               131328    
_________________________________________________________________
dense_82 (Dense)             (None, 256)               65792     
_________________________________________________________________
dense_83 (Dense)             (None, 128)               32896     
_________________________________________________________________
dense_84 (Dense)             (None, 1)                 129       
Total params: 341,249
Trainable params: 341,249
Non-trainable params: 0
_________________________________________________________________
None


In [66]:
#Entreno el modelo con 150 épocas y batch_size de 256
modelo2_start = time.time()
modelo2.fit(X_train,
                  y_train,
                  epochs=150,
                  batch_size=256,
                  validation_split = 0.1  #Con esto se separa una porción del conjunto de entratamiento y se usa como
                                          #validación, con esa porción no se entrena
                  )
modelo2_end = time.time()
print(f"Tiempo de ejecución: {round((modelo2_end - modelo2_start)/60,1)} minutes")

Train on 8154 samples, validate on 906 samples
Epoch 1/150
8154/8154 [==============================] - 2s 213us/step - loss: 74.4532 - mean_squared_error: 1.7511 - val_loss: 61.4748 - val_mean_squared_error: 0.3123
Epoch 2/150
8154/8154 [==============================] - 0s 20us/step - loss: 50.5951 - mean_squared_error: 0.2326 - val_loss: 39.6972 - val_mean_squared_error: 0.1898
Epoch 3/150
8154/8154 [==============================] - 0s 18us/step - loss: 31.4935 - mean_squared_error: 0.1556 - val_loss: 23.9814 - val_mean_squared_error: 0.1478
Epoch 4/150
8154/8154 [==============================] - 0s 19us/step - loss: 19.2437 - mean_squared_error: 0.1320 - val_loss: 15.1402 - val_mean_squared_error: 0.1324
Epoch 5/150
8154/8154 [==============================] - 0s 20us/step - loss: 12.5760 - mean_squared_error: 0.1242 - val_loss: 10.2963 - val_mean_squared_error: 0.1282
Epoch 6/150
8154/8154 [==============================] - 0s 21us/step - loss: 8.7636 - mean_squared_error: 0.122

In [67]:
#Evaluo el modelo con las muestras de test

y_test_pred = modelo2.predict(X_test)
y_train_pred = modelo2.predict(X_train)
print("Training MSE:", round(mean_squared_error(y_train, y_train_pred),4))
print("Validation MSE:", round(mean_squared_error(y_test, y_test_pred),4))
print("\nTraining r2:", round(r2_score(y_train, y_train_pred),4))
print("Validation r2:", round(r2_score(y_test, y_test_pred),4))

Training MSE: 0.1034
Validation MSE: 0.1006

Training r2: 0.708
Validation r2: 0.7152


Este modelo se comporta mucho mejor. Se ha eliminado el problema de overfitting que teníamos antes y ha mejorado también r2

## 4.3. Modelo 3: Red neuronal con 4 capas, regularización Droput 0.5 y más épocas
Voy a probar con la regularización Dropout=0.5


In [53]:
# Construyo el modelo
modelo3 = models.Sequential()
modelo3.add(layers.Dropout(0.5, input_shape=(X_train.shape[1],)))
modelo3.add(layers.Dense(512, activation='relu'))
modelo3.add(layers.Dropout(0.5))
modelo3.add(layers.Dense(256, activation='relu'))
modelo3.add(layers.Dropout(0.5))
modelo3.add(layers.Dense(256, activation='relu'))
modelo3.add(layers.Dropout(0.5))
modelo3.add(layers.Dense(128, activation='relu'))
modelo3.add(layers.Dropout(0.5))
modelo3.add(layers.Dense(1, activation='linear'))

# Compilo el modelo
modelo3.compile(loss='mean_squared_error',
            optimizer='adam',
            metrics=['mean_squared_error'])

# Model summary
print(modelo3.summary())

Model: "sequential_13"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dropout_51 (Dropout)         (None, 216)               0         
_________________________________________________________________
dense_60 (Dense)             (None, 512)               111104    
_________________________________________________________________
dropout_52 (Dropout)         (None, 512)               0         
_________________________________________________________________
dense_61 (Dense)             (None, 256)               131328    
_________________________________________________________________
dropout_53 (Dropout)         (None, 256)               0         
_________________________________________________________________
dense_62 (Dense)             (None, 256)               65792     
_________________________________________________________________
dropout_54 (Dropout)         (None, 256)             

In [54]:
#Entreno el modelo con 260 épocas y batch_size de 256
modelo3_start = time.time()
modelo3.fit(X_train,
                  y_train,
                  epochs=260,
                  batch_size=256,
                  validation_split = 0.1  #Con esto se separa una porción del conjunto de entratamiento y se usa como
                                          #validación, con esa porción no se entrena
                  )
modelo3_end = time.time()
print(f"Tiempo de ejecución: {round((modelo3_end - modelo3_start)/60,1)} minutes")

Train on 8154 samples, validate on 906 samples
Epoch 1/260
8154/8154 [==============================] - 2s 225us/step - loss: 5.3962 - mean_squared_error: 5.3962 - val_loss: 8.4412 - val_mean_squared_error: 8.4412
Epoch 2/260
8154/8154 [==============================] - 0s 22us/step - loss: 2.3019 - mean_squared_error: 2.3019 - val_loss: 6.7733 - val_mean_squared_error: 6.7733
Epoch 3/260
8154/8154 [==============================] - 0s 21us/step - loss: 1.9900 - mean_squared_error: 1.9900 - val_loss: 7.7305 - val_mean_squared_error: 7.7305
Epoch 4/260
8154/8154 [==============================] - 0s 21us/step - loss: 1.6772 - mean_squared_error: 1.6772 - val_loss: 7.6583 - val_mean_squared_error: 7.6583
Epoch 5/260
8154/8154 [==============================] - 0s 19us/step - loss: 1.4228 - mean_squared_error: 1.4228 - val_loss: 6.5603 - val_mean_squared_error: 6.5603
Epoch 6/260
8154/8154 [==============================] - 0s 19us/step - loss: 1.3079 - mean_squared_error: 1.3079 - val_lo

In [55]:
#Evaluo el modelo con las muestras de test

y_test_pred = modelo3.predict(X_test)
y_train_pred = modelo3.predict(X_train)
print("Training MSE:", round(mean_squared_error(y_train, y_train_pred),4))
print("Validation MSE:", round(mean_squared_error(y_test, y_test_pred),4))
print("\nTraining r2:", round(r2_score(y_train, y_train_pred),4))
print("Validation r2:", round(r2_score(y_test, y_test_pred),4))

Training MSE: 0.1151
Validation MSE: 0.1305

Training r2: 0.6748
Validation r2: 0.6307


El comportamiento de este modelo es algo peor que el de la regularización L1

## 4.4. Modelo 4: Red neuronal con 4 capas, regularización Dropout 0.2 y mayor batch size
Voy a probar con la regularización Droput al 0.2 y un aumento del batch_size a 512

In [74]:
# Construyo el modelo
modelo4 = models.Sequential()
modelo4.add(layers.Dropout(0.2, input_shape=(X_train.shape[1],)))
modelo4.add(layers.Dense(512, activation='relu'))
modelo4.add(layers.Dropout(0.2))
modelo4.add(layers.Dense(256, activation='relu'))
modelo4.add(layers.Dropout(0.2))
modelo4.add(layers.Dense(256, activation='relu'))
modelo4.add(layers.Dropout(0.2))
modelo4.add(layers.Dense(128, activation='relu'))
modelo4.add(layers.Dropout(0.2))
modelo4.add(layers.Dense(1, activation='linear'))

# Compiling the model
modelo4.compile(loss='mean_squared_error',
            optimizer='adam',
            metrics=['mean_squared_error'])

# Model summary
print(modelo4.summary())

Model: "sequential_20"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dropout_66 (Dropout)         (None, 216)               0         
_________________________________________________________________
dense_95 (Dense)             (None, 512)               111104    
_________________________________________________________________
dropout_67 (Dropout)         (None, 512)               0         
_________________________________________________________________
dense_96 (Dense)             (None, 256)               131328    
_________________________________________________________________
dropout_68 (Dropout)         (None, 256)               0         
_________________________________________________________________
dense_97 (Dense)             (None, 256)               65792     
_________________________________________________________________
dropout_69 (Dropout)         (None, 256)             

In [75]:
#Entreno el modelo con 150 épocas y batch_size de 512
modelo4_start = time.time()
modelo4.fit(X_train,
                  y_train,
                  epochs=300,
                  batch_size=512,
                  validation_split = 0.1  #Con esto se separa una porción del conjunto de entratamiento y se usa como
                                          #validación, con esa porción no se entrena
                  )
modelo4_end = time.time()
print(f"Tiempo de ejecución: {round((modelo4_end - modelo4_start)/60,1)} minutes")

Train on 8154 samples, validate on 906 samples
Epoch 1/300
8154/8154 [==============================] - 2s 250us/step - loss: 3.8492 - mean_squared_error: 3.8492 - val_loss: 0.8725 - val_mean_squared_error: 0.8725
Epoch 2/300
8154/8154 [==============================] - 0s 11us/step - loss: 1.1134 - mean_squared_error: 1.1134 - val_loss: 0.5969 - val_mean_squared_error: 0.5969
Epoch 3/300
8154/8154 [==============================] - 0s 11us/step - loss: 0.7738 - mean_squared_error: 0.7738 - val_loss: 0.7688 - val_mean_squared_error: 0.7688
Epoch 4/300
8154/8154 [==============================] - 0s 11us/step - loss: 0.6325 - mean_squared_error: 0.6325 - val_loss: 1.0120 - val_mean_squared_error: 1.0120
Epoch 5/300
8154/8154 [==============================] - 0s 12us/step - loss: 0.5431 - mean_squared_error: 0.5431 - val_loss: 0.9258 - val_mean_squared_error: 0.9258
Epoch 6/300
8154/8154 [==============================] - 0s 11us/step - loss: 0.5108 - mean_squared_error: 0.5108 - val_lo

In [76]:
#Evaluo el modelo con las muestras de test

y_test_pred = modelo4.predict(X_test)
y_train_pred = modelo4.predict(X_train)
print("Training MSE:", round(mean_squared_error(y_train, y_train_pred),4))
print("Validation MSE:", round(mean_squared_error(y_test, y_test_pred),4))
print("\nTraining r2:", round(r2_score(y_train, y_train_pred),4))
print("Validation r2:", round(r2_score(y_test, y_test_pred),4))

Training MSE: 0.0404
Validation MSE: 0.1245

Training r2: 0.8859
Validation r2: 0.6477


Esto ha emperado el modelo.Seguimos quedándonos por tanto con el modelo 2 y la regularización L1

## 4.5. Modelo 5: Red neuronal con 4 capas, regularización L1 y eliminación de más características correladas

En el limpiado del dataset vi que existía correlación entre las características review pero no las eliminé en ese momento. Voy a probar si al eliminarlas y usar el modelo de regularización L1 conseguimos mejor resultado

In [0]:
all_review = list(X_train.columns[X_train.columns.str.startswith("review_scores")])
review_to_keep = list(X_train.columns[X_train.columns.str.startswith("review_scores_rating")])
review_to_drop = [x for x in all_review if x not in review_to_keep]

X_train_short = X_train.drop(review_to_drop, axis=1)

#Misma transformación para test
X_test_short = X_test.drop(review_to_drop, axis=1)

In [96]:
# Construyo el modelo igual que el modelo2
modelo5 = models.Sequential()
modelo5.add(layers.Dense(512, input_shape=(X_train_short.shape[1],), kernel_regularizer=regularizers.l1(0.005), activation='relu'))
modelo5.add(layers.Dense(256, kernel_regularizer=regularizers.l1(0.005), activation='relu'))
modelo5.add(layers.Dense(256, kernel_regularizer=regularizers.l1(0.005), activation='relu'))
modelo5.add(layers.Dense(128, kernel_regularizer=regularizers.l1(0.005), activation='relu'))
modelo5.add(layers.Dense(1, activation='linear'))

# Compilo el modelo
modelo5.compile(loss='mean_squared_error',
                optimizer='adam',
                metrics=['mean_squared_error'])

# Model summary
print(modelo5.summary())

Model: "sequential_27"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_131 (Dense)            (None, 512)               95744     
_________________________________________________________________
dense_132 (Dense)            (None, 256)               131328    
_________________________________________________________________
dense_133 (Dense)            (None, 256)               65792     
_________________________________________________________________
dense_134 (Dense)            (None, 128)               32896     
_________________________________________________________________
dense_135 (Dense)            (None, 1)                 129       
Total params: 325,889
Trainable params: 325,889
Non-trainable params: 0
_________________________________________________________________
None


In [97]:
#Entreno el modelo con 150 épocas y batch_size de 256
modelo5_start = time.time()
modelo5.fit(X_train_short,
                  y_train,
                  epochs=150,
                  batch_size=256,
                  validation_split = 0.1  #Con esto se separa una porción del conjunto de entratamiento y se usa como
                                          #validación, con esa porción no se entrena
                  )
modelo5_end = time.time()
print(f"Tiempo de ejecución: {round((modelo5_end - modelo5_start)/60,1)} minutes")

Train on 8154 samples, validate on 906 samples
Epoch 1/150
8154/8154 [==============================] - 3s 324us/step - loss: 70.5880 - mean_squared_error: 1.7149 - val_loss: 56.9173 - val_mean_squared_error: 0.3800
Epoch 2/150
8154/8154 [==============================] - 0s 20us/step - loss: 45.6667 - mean_squared_error: 0.2261 - val_loss: 34.6640 - val_mean_squared_error: 0.2068
Epoch 3/150
8154/8154 [==============================] - 0s 19us/step - loss: 26.6554 - mean_squared_error: 0.1491 - val_loss: 19.5208 - val_mean_squared_error: 0.1435
Epoch 4/150
8154/8154 [==============================] - 0s 21us/step - loss: 15.2713 - mean_squared_error: 0.1300 - val_loss: 11.7000 - val_mean_squared_error: 0.1255
Epoch 5/150
8154/8154 [==============================] - 0s 20us/step - loss: 9.5946 - mean_squared_error: 0.1270 - val_loss: 7.7378 - val_mean_squared_error: 0.1221
Epoch 6/150
8154/8154 [==============================] - 0s 18us/step - loss: 6.5456 - mean_squared_error: 0.1237 

In [98]:
#Evaluo el modelo con las muestras de test

y_test_pred = modelo5.predict(X_test_short)
y_train_pred = modelo5.predict(X_train_short)
print("Training MSE:", round(mean_squared_error(y_train, y_train_pred),4))
print("Validation MSE:", round(mean_squared_error(y_test, y_test_pred),4))
print("\nTraining r2:", round(r2_score(y_train, y_train_pred),4))
print("Validation r2:", round(r2_score(y_test, y_test_pred),4))

Training MSE: 0.1071
Validation MSE: 0.1039

Training r2: 0.6974
Validation r2: 0.7061


In [0]:
import pickle
#Guardo este modelo por si es el bueno defintivo
path_df = ruta + 'model/RedNeuronal.sav'
pickle.dump(modelo5, open(path_df, 'wb'))

Conseguimos resultados similares al del modelo 2 pero he eliminado 30 características con lo que la complejidad es menor por lo que nos quedamos de momento con esta opción

## 4.6. Modelo 6: Red neuronal con 5 capas, regularización L1 y eliminación de características correladas

Viendo que el mejor modelo es modelo 5 voy a intentar mejorar sus prestaciones añadiendo una capa densa más

In [81]:
# Construyo el modelo igual que el modelo2
modelo6 = models.Sequential()
modelo6.add(layers.Dense(1024, input_shape=(X_train_short.shape[1],), kernel_regularizer=regularizers.l1(0.005), activation='relu'))
modelo6.add(layers.Dense(512, kernel_regularizer=regularizers.l1(0.005), activation='relu'))
modelo6.add(layers.Dense(256, kernel_regularizer=regularizers.l1(0.005), activation='relu'))
modelo6.add(layers.Dense(256, kernel_regularizer=regularizers.l1(0.005), activation='relu'))
modelo6.add(layers.Dense(128, kernel_regularizer=regularizers.l1(0.005), activation='relu'))
modelo6.add(layers.Dense(1, activation='linear'))

# Compilo el modelo
modelo6.compile(loss='mean_squared_error',
                optimizer='adam',
                metrics=['mean_squared_error'])

# Model summary
print(modelo6.summary())

Model: "sequential_22"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_105 (Dense)            (None, 1024)              191488    
_________________________________________________________________
dense_106 (Dense)            (None, 512)               524800    
_________________________________________________________________
dense_107 (Dense)            (None, 256)               131328    
_________________________________________________________________
dense_108 (Dense)            (None, 256)               65792     
_________________________________________________________________
dense_109 (Dense)            (None, 128)               32896     
_________________________________________________________________
dense_110 (Dense)            (None, 1)                 129       
Total params: 946,433
Trainable params: 946,433
Non-trainable params: 0
_______________________________________________

In [82]:
#Entreno el modelo con 150 épocas y batch_size de 256
modelo6_start = time.time()
modelo6.fit(X_train_short,
                  y_train,
                  epochs=150,
                  batch_size=256,
                  validation_split = 0.1  #Con esto se separa una porción del conjunto de entratamiento y se usa como
                                          #validación, con esa porción no se entrena
                  )
modelo6_end = time.time()
print(f"Tiempo de ejecución: {round((modelo6_end - modelo6_start)/60,1)} minutes")

Train on 8154 samples, validate on 906 samples
Epoch 1/150
8154/8154 [==============================] - 2s 296us/step - loss: 139.7885 - mean_squared_error: 1.7528 - val_loss: 100.1704 - val_mean_squared_error: 0.3546
Epoch 2/150
8154/8154 [==============================] - 0s 23us/step - loss: 71.0743 - mean_squared_error: 0.2400 - val_loss: 45.6358 - val_mean_squared_error: 0.1769
Epoch 3/150
8154/8154 [==============================] - 0s 24us/step - loss: 32.3395 - mean_squared_error: 0.1490 - val_loss: 22.0344 - val_mean_squared_error: 0.1535
Epoch 4/150
8154/8154 [==============================] - 0s 23us/step - loss: 16.9240 - mean_squared_error: 0.1313 - val_loss: 12.8568 - val_mean_squared_error: 0.1244
Epoch 5/150
8154/8154 [==============================] - 0s 22us/step - loss: 10.6505 - mean_squared_error: 0.1242 - val_loss: 8.7727 - val_mean_squared_error: 0.1189
Epoch 6/150
8154/8154 [==============================] - 0s 24us/step - loss: 7.6375 - mean_squared_error: 0.12

In [83]:
#Evaluo el modelo con las muestras de test

y_test_pred = modelo6.predict(X_test_short)
y_train_pred = modelo6.predict(X_train_short)
print("Training MSE:", round(mean_squared_error(y_train, y_train_pred),4))
print("Validation MSE:", round(mean_squared_error(y_test, y_test_pred),4))
print("\nTraining r2:", round(r2_score(y_train, y_train_pred),4))
print("Validation r2:", round(r2_score(y_test, y_test_pred),4))

Training MSE: 0.1064
Validation MSE: 0.1031

Training r2: 0.6995
Validation r2: 0.7083


Se consigue muy poca mejora para el aumento de complejidad

## 4.7. Conclusión. Modelo Red Neuronal con datos solo numéricos

Tras las diversas pruebas se selecciona el modelo 5 con mejor modelo: Red neuronal con 4 capas, regularización L1, optimización de Adam y eliminación de características review correladas:

Training r2: 0.6974

Validation r2: 0.7061

